### ISSUE TITLE
switch to dune instead of ocamlbuild

### ISSUE THREAD

**Issue by UnixJunkie** [ocaml-batteries-team/batteries-included#738]
if it parallelizes the build and works on windows, that would be quite nice.

**Comment by gasche** [comment]
I believe that ocamlbuild works on Windows but yes, feel free to try jbuilder -- we do need to make sure that it does also work on Windows, I think.

**Comment by UnixJunkie** [comment]
we should really give it a try. I guess it would parallelize the build quite well.

**Comment by c-cube** [comment]
Just my 2 cents: on containers, dune makes the build fast and parallelizes well. It also allows qtest to be used with a few lines in jbuild files.

**Comment by gasche** [comment]
I'm sure it would be nice to have a dune-based build system for Batteries, and it should not be too complicated -- we're not doing complex build stuff. The question is rather of who is motivated to implement it. (It might be simpler to move to `cppo` than to preserve the `prefilter` stuff, which is probably a good idea.)

(What's the bound on OCaml versions that comes with adopting dune? Should we keep an ocamlbuild or Makefile on the side for older OCaml versions?)

**Comment by c-cube** [comment]
the bound is >= 4.02.3 in practice, I think. It's kind of annoying for me too, but it's a tradeoff to make… :/

**Comment by UnixJunkie** [comment]
I'd also like that we move to cppo. I'd also like that all files are .ml/.mli. I don't see why some files have to be called .mlv/.mliv. It's just annoying when you edit the wrong file, and its not uniform.
The build system should know when to preprocess a file or not.
And, if all files are preprocessed, a preprocessor is something that is supposed to be lightning fast.

**Comment by UnixJunkie** [comment]
If we switch to jbuilder, this may send a signal to some people that at least part of the open-source ocaml community is seriously considering jbuilder, and it may incent more people to switch to it.
I think it is very positive for the ocaml ecosystem if the community starts to move "en masse" to such a build system. The previous bazaar of build systems in ocaml was a problem for newcomers, and the bad taste of the build system that was shipping with the compiler for years was also problematic (though this is maybe just my personal opinion).

**Comment by UnixJunkie** [comment]
Some venerable OCaml programmers have not switched yet...

**Comment by UnixJunkie** [comment]
I haven't switched all my projects yet also, but this might happen over time.

**Comment by XVilka** [comment]
Oasis is being deprecated and switch to [Dune](https://dune.build) is recommended, see the discussion here: https://discuss.ocaml.org/t/will-you-use-oasis-dune-migration-tool/4213

**Comment by UnixJunkie** [comment]
> Oasis is being deprecated and switch to [Dune](https://dune.build) is recommended

Contributions are warmly welcome. :)

**Comment by UnixJunkie** [comment]
@gasche do you still have some dune-switch related things to contribute?

**Comment by fccm** [comment]
`make` is also able to parallelise the build.

**Comment by UnixJunkie** [comment]
And maintaining makefiles is such a pleasure...

**Comment by gasche** [comment]
> @gasche do you still have some dune-switch related things to contribute?

Sorry for not being very reactive. I'm swamped with work right now, probably until at least the new year.

I don't remember which parts of my moving-to-dune effort I had submitted, and which part I haven't. My most recent dune branch is available at my fork, in branch [dune-wip](https://github.com/gasche/batteries-included/commits/dune-wip).

**Comment by UnixJunkie** [comment]
Was your branch able to compile with dune or is it still a work in progress?

**Comment by UnixJunkie** [comment]
Please send your full branch as a PR.
I'll try to merge things after the still-to-be-done NEXT_RELEASE.

**Comment by UnixJunkie** [comment]
@sim642 do you want to give a stab at this issue?
I think you were the biggest contributor to the latest release.
There has already been some work done by @gasche that was merged.
So maybe little work is required to finish this one.

**Comment by sim642** [comment]
I experimented with the existing dune stuff a couple of months ago, so sure, I can have a look at it again sometime.

**Comment by UnixJunkie** [comment]
Congratulations to @gasche for the initial push, and to @sim642 for having finally killed this issue.

**Comment by sim642** [comment]
I think `dune install` still isn't entirely what it should be. It was mentioned somewhere that it probably makes sense to move the special toploop stuff to a separate `batteries.top` findlib library, which is what many other packages do. I'll probably have a look at that sometime.

It will be hard, if not impossible, to maintain compatible ocamlbuild and dune setups in parallel, so it would still make sense to strip ocamlbuild out. It would also simplify some dune stuff, which currently is a bit hacky.

**Comment by UnixJunkie** [comment]
If dune can do what the Makefile-using-ocamlbuild can do, using ocaml >= 4.02.3, I think we can get rid of ocamlbuild.
The critical things I am aware of are: 'make', 'make test', 'make install', 'make doc'.
